types/parsimmon/parsimmon-tests.ts

```diff
@@ -191,3 +191,55 @@ strArrPar = P.sepBy1(P.string('foo'), P.string('bar'));
 strPar = P.test((a: string) => false);
 
 strPar = P.takeWhile((a: string) => true);
+
+// --  --  --  --  --  --  --  --  --  --  --  --  --
+
+let language: Language;
+
+language = P.createLanguage({
+	SomeRule: r => P.alt(P.string(""), r.AnotherRule),
+	AnotherRule: () => P.string(""),
+});
+
+// $ExpectType Parser<any>
```

Comment:
@bvanreeven Is there a way to better type these? Ie. this rule should ideally be `Parser<string>`.

Reply:
As far as I can tell, this is not possible without introducing additional type parameters: one for the `rules` parameter, and one for the return type.

The return type of `createLanguage` should be an object that has the same properties as the `rules` parameter object, but the types of those properties should match the return types of the functions that are passed as properties of the `rules` parameter object. If that makes sense.

I tried an approach that has only one type parameter and returns a mapped type of the parameter, but I got stuck here:

```typescript
function createLanguage<TRule extends Rule>(rules: TRule): { [K in keyof TRule]: <return type of>TRule[K] };
```

As far as I know, it is currently not possible to statically determine the return type of a function type.

Reply:
@bcherny I added some documentation on how to use the type parameter in `createLanguage` to make a language specification typed.